pkg/placement/placement-decision.go

```diff
@@ -0,0 +1,144 @@
+	obj, err := c.getObjectFromKey(key)
+	if err != nil {
+		// The resource doesn't exist, needs to be created
+		obj = util.EmptyUnstructuredObjectFromKey(key)
+	}
+
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("failed to convert runtime.Object to unstructured.Unstructured")
+	}
+
+	placementDecision, err := unstructuredObjectToPlacementDecision(unstructuredObj)
+	if err != nil {
+		return fmt.Errorf("failed to convert from Unstructured to PlacementDecision: %w", err)
+	}
+
+	// placement decision name matches that of the placement 1:1, therefore its NamespacedName is the same.
+	placementIdentifier := placementDecision.GetName()
+
+	// generate placement decision spec from resolver
+	generatedPlacementDecisionSpec, err := c.placementResolver.GeneratePlacementDecision(placementIdentifier)
+	if err != nil {
+		return fmt.Errorf("failed to get placement decision spec: %w", err)
+	}
+
+	// calculate if the resolved decision is different from the current one
+	if !c.placementResolver.ComparePlacementDecision(placementIdentifier, &placementDecision.Spec) {
+		// update the placement decision object in the cluster by updating spec
+		if err = c.updateOrCreatePlacementDecision(placementDecision, generatedPlacementDecisionSpec); err != nil {
+			return fmt.Errorf("failed to update or create placement decision: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// updateOrCreatePlacementDecision updates or creates a placement-decision object in the cluster.
+// If the object already exists, it is updated. Otherwise, it is created.
+func (c *Controller) updateOrCreatePlacementDecision(pd *v1alpha1.PlacementDecision,
+	placementDecisionSpec *v1alpha1.PlacementDecisionSpec) error {
+	// use the passed placement decision and set its spec
+	pd.Spec = *placementDecisionSpec
+
+	// set owner reference
+	//	get owner placement object
+	placement, err := c.getPlacementByName(pd.GetName())
+	if err != nil {
+		return fmt.Errorf("failed to get placement %s: %w", pd.GetName(), err)
+	}
+
+	setOwnerReference(placement, pd)
+
+	// update or create placement decision
+	unstructuredPlacementDecision, err := placementDecisionToUnstructuredObject(pd)
+	if err != nil {
+		return fmt.Errorf("failed to update or create placement decision: %w", err)
+	}
+
+	_, err = c.dynamicClient.Resource(schema.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  pd.GetObjectKind().GroupVersionKind().Version,
+		Resource: util.PlacementDecisionResource,
+	}).Update(context.Background(), unstructuredPlacementDecision, metav1.UpdateOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_, err = c.dynamicClient.Resource(schema.GroupVersionResource{
+				Group:    v1alpha1.SchemeGroupVersion.Group,
+				Version:  pd.GetObjectKind().GroupVersionKind().Version,
+				Resource: util.PlacementDecisionResource,
+			}).Create(context.Background(), unstructuredPlacementDecision, metav1.CreateOptions{})
+			if err != nil {
```

Comment:
In the success case it would be better to log a "Created" message and return than to fall through to logging an "Updated" message.

Reply:
Are you suggesting I only log creation?

Reply:
Misunderstood previously (and previous comment on this I guess), fixed